api/src/repositories/observation-repository.ts

```diff
@@ -0,0 +1,164 @@
+import moment from 'moment';
+import SQL from 'sql-template-strings';
+import { z } from 'zod';
+import { getKnex } from '../database/db';
+import { BaseRepository } from './base-repository';
+
+/**
+ * Interface reflecting survey observations retrieved from the database
+ */
+export const ObservationRecord = z.object({
+  survey_observation_id: z.number(),
+  survey_id: z.number(),
+  wldtaxonomic_units_id: z.number(),
+  latitude: z.number(),
+  longitude: z.number(),
+  count: z.number(),
+  observation_time: z.string(),
+  observation_date: z.string(),
+  create_date: z.string(),
+  create_user: z.number(),
+  update_date: z.string().nullable(),
+  update_user: z.number().nullable(),
+  revision_count: z.number()
+});
+
+export type ObservationRecord = z.infer<typeof ObservationRecord>;
+
+/**
+ * Interface reflecting survey observations that are being inserted into the database
+ */
+export type InsertObservation = Pick<
+  ObservationRecord,
+  'survey_id' | 'wldtaxonomic_units_id' | 'latitude' | 'longitude' | 'count' | 'observation_date' | 'observation_time'
+>;
+
+/**
+ * Interface reflecting survey observations that are being updated in the database
+ */
+export type UpdateObservation = Pick<
+  ObservationRecord,
+  | 'survey_observation_id'
+  | 'wldtaxonomic_units_id'
+  | 'latitude'
+  | 'longitude'
+  | 'count'
+  | 'observation_date'
+  | 'observation_time'
+>;
+
+export class ObservationRepository extends BaseRepository {
+  /**
+   * Deletes all survey observation records associated with the given survey, except
+   * for records whose ID belongs to the given array, then returns the count of
+   * affected rows.
+   *
+   * @param {number} surveyId
+   * @param {number[]} retainedObservationIds Observation records to retain (not be deleted)
+   * @return {*}  {Promise<number>}
+   * @memberof ObservationRepository
+   */
+  async deleteObservationsNotInArray(surveyId: number, retainedObservationIds: number[]): Promise<number> {
+    const sqlStatement = SQL`
+      DELETE FROM
+        survey_observation
+      WHERE
+        survey_id = ${surveyId}
+    `;
+
+    if (retainedObservationIds.length) {
+      sqlStatement.append(`
+        AND
+          survey_observation_id
+        NOT IN
+          (${retainedObservationIds.join(',')})
+      `);
+    }
+
+    sqlStatement.append(';');
+
+    const response = await this.connection.sql(sqlStatement);
+
+    return response.rowCount;
+  }
+
+  /**
+   * Performs an upsert for all observation records belonging to the given survey, then
+   * returns the updated rows
+   *
+   * @param {number} surveyId
+   * @param {((InsertObservation | UpdateObservation)[])} observations
+   * @return {*}  {Promise<ObservationRecord[]>}
+   * @memberof ObservationRepository
+   */
+  async insertUpdateSurveyObservations(
```

Comment:
I spent a lot of time workshopping this with Knex `merge` and could not get it to work, chiefly because entering a raw SQL allows me to `OVERRIDING SYSTEM VALUE` when merging the `survey_observation_id`